### ISSUE TITLE
Getting issue in grpc Bi-directional method call

### ISSUE THREAD

**Issue by NeerajThapliyal** [phiSgr/gatling-grpc#1]
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at io.gatling.mojo.MainWithArgsInFile.runMain(MainWithArgsInFile.java:50)
	at io.gatling.mojo.MainWithArgsInFile.main(MainWithArgsInFile.java:33)
Caused by: java.lang.AssertionError: assertion failed
	at scala.Predef$.assert(Predef.scala:204)
	at com.github.phisgr.gatling.grpc.GrpcDsl$Call.rpc(GrpcDsl.scala:26)
	at BiDirectionalQueryTest.<init>(BiDirectionalQueryTest.scala:48)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at io.gatling.app.Runner.run0(Runner.scala:79)
	at io.gatling.app.Runner.run(Runner.scala:64)
	at io.gatling.app.Gatling$.start(Gatling.scala:59)
	at io.gatling.app.Gatling$.fromArgs(Gatling.scala:43)
	at io.gatling.app.Gatling$.main(Gatling.scala:35)
	at io.gatling.app.Gatling.main(Gatling.scala)
	... 6 more

**Comment by NeerajThapliyal** [comment]
@phiSgr thanks for reply
actually i am not too much familiar with scala.
i want to test  Bi-directional grpc method using gatling.
can you provide me way how i can integrate it.

**Comment by phiSgr** [comment]
@NeerajThapliyal it would be great if you can let me know how you want the bi-di stream to be tested.

For example, I implemented a (server-side) streaming rpc to return the latest version of a value and stream the new values to come. 
Like the [BehaviorSubject](http://reactivex.io/RxJava/javadoc/rx/subjects/BehaviorSubject.html) in Rx.
![](https://raw.githubusercontent.com/wiki/ReactiveX/RxJava/images/rx-operators/S.BehaviorSubject.png)

Then in the stress test I measured the time it takes for the first value to arrive.
```scala
grpc("Stream Example")
  .service(StreamServiceGrpc.stub)
  .rpc(service => { _: StreamReq =>
    val p = Promise[StreamRes]()
    service.s(new ClientResponseObserver[StreamReq, StreamRes] {
      var requestStream: ClientCallStreamObserver[StreamReq] = _

      override def beforeStart(requestStream: ClientCallStreamObserver[StreamReq]): Unit = {
        this.requestStream = requestStream
      }

      override def onNext(value: StreamRes): Unit = {
        p.trySuccess(value)
        requestStream.cancel("bye", null)
      }

      override def onError(t: Throwable): Unit = {
        p.tryFailure(t)
      }

      override def onCompleted(): Unit = {}
    })
    p.future
  })(StreamReq())
```

Maybe your intended test cannot be done with the current version of the library, but let me know anyway.
The future design for streaming support can definitely benefit from more use cases.

**Comment by NeerajThapliyal** [comment]
@phiSgr  thanks for reply 
i have tried this but i am stuck on second line. i am have no idea what should i put on second line.
here my code..

import _root_.io.gatling.core.scenario.Simulation
import ch.qos.logback.classic.{Level, LoggerContext}
import io.gatling.core.Predef._
import io.gatling.http.Predef._
import org.slf4j.LoggerFactory
import com.github.phisgr.gatling.grpc.Predef._
import com.github.phisgr.gatling.pb._
import com.github.phisgr.gatling.util._
// stringToExpression is hidden because we have $ in GrpcDsl
import io.gatling.core.Predef.{stringToExpression => _, _}
import io.gatling.core.session.Expression
import io.grpc.{ManagedChannelBuilder, Status}
import io.grpc.ManagedChannel
import io.grpc.stub._

import javax.net.ssl.SSLException;
import java.io.File;
import com.flair.bi.messages.QueryServiceGrpc;
import io.grpc.Channel

import io.grpc.netty.shaded.io.grpc.netty.GrpcSslContexts;
import io.grpc.netty.shaded.io.grpc.netty.NettyChannelBuilder;
import com.flair.bi.messages.Query;
import com.flair.bi.messages.QueryResponse;

import scala.concurrent.duration._
/**
 * Performance test for the Connection entity.
 */
class BiDirectionalQueryTest extends Simulation {

  
   val grpcConf = grpc(NettyChannelBuilder.forAddress("localhost", 6565).sslContext(GrpcSslContexts.forClient().trustManager(new File("src/main/resources/ssl/grpc/server.crt")).build()));
  val managedChannel: Channel  = NettyChannelBuilder.forAddress("localhost", 6565).sslContext(GrpcSslContexts.forClient().trustManager(new File("src/main/resources/ssl/grpc/server.crt")).build()).build();
  
//  val queryStub: QueryServiceGrpc.QueryServiceBlockingStub  = QueryServiceGrpc.newBlockingStub(managedChannel);

  
   val  query: Query = Query.newBuilder()
                .setQueryId("1")
                .setUserId("34fsdfgd")
                .setSourceId("1715917d-fff8-44a1-af02-ee2cd41a3609")
                .setSource("Transactions local")
                .addFields("state")
                .addFields("city")
                .setLimit(10)
                .build();
   
  
 val s =  grpc("Bi-Directional Query Test")
  .service(QueryServiceGrpc.getGetDataStreamMethod)
  .rpc(service => { _: StreamReq =>
    val p = Promise[StreamRes]()
    service.s(new ClientResponseObserver[StreamReq, StreamRes] {
      var requestStream: ClientCallStreamObserver[StreamReq] = _

      override def beforeStart(requestStream: ClientCallStreamObserver[StreamReq]): Unit = {
        this.requestStream = requestStream
      }

      override def onNext(value: StreamRes): Unit = {
        p.trySuccess(value)
        requestStream.request(query);
       // requestStream.cancel("bye", null)
      }

      override def onError(t: Throwable): Unit = {
        p.tryFailure(t)
      }

      override def onCompleted(): Unit = {}
    })
    p.future
  })(StreamReq())
    

  setUp(
    s.inject(atOnceUsers(5))
  ).protocols(grpcConf)
}

**Comment by phiSgr** [comment]
It fails to compile because in the old API, `.service` expects a method reference that returns a service from a `Channel`.

For that to compile you have to pass in `QueryServiceGrpc.stub` instead of `QueryServiceGrpc.getGetDataStreamMethod`, which returns a method descriptor.

**Comment by rene-m-hernandez** [comment]
Use case I would like to be able to test for: time between responses sent by server.

I have a scenario where a client establishes a bi-directional streaming gRPC channel with the server. The client will send an initial request and the server will return an initial response. The time between the initial request and the initial response is somewhat important, but I am more interested in what happens after the initial response.

After the initial response, the channel will remain open indefinitely. The server will then send back responses as it desires. These responses are triggered by an external event. Ideally, I would _love_ to be able to measure the time between the external event and the client receiving a new response from the server (triggered by the external event). I'm not sure how feasible that is; alternatively, if I could simply measure the time between responses sent by the server, I think that would be good enough.

Unsure if Gatling is the proper tool for this. I've had a quick look at ghz and fortio, but I'm unsure if they suit my needs either.

Maybe something like this: https://gatling.io/docs/2.2/http/sse/#server-messages-checks

**Comment by phiSgr** [comment]
Thanks for commenting @rene-m-hernandez. I had something similar in mind, largely inspired by the chatroom example referenced in [documentation on WebSocket in Gatling](https://gatling.io/docs/2.2/http/websocket/#example).

If the external event (e.g. a chat message from another user) has a timestamp, it is more than feasible to get the difference in the timestamps when the response is received.
The value is one-way latency here. It is not always sensible to compare timestamps from different computers. But in this use case this is probably the best we can have.

For a estimate of how accurately clocks are synchronized, A quick search on Google provided me [this](https://serverfault.com/questions/875037/how-accurately-are-aws-clocks-synchronized-between-different-aws-regions/).

**Comment by ddemonaz** [comment]
good afternoon! Tell me if streaming support has appeared? Are there any ways to do this on the current version of the library?

**Comment by phiSgr** [comment]
I have something that is not yet out because I do not have enough time to polish it, and that I do not see enough people needing it.

**Comment by ddemonaz** [comment]
grpc("Stream Example")
  .service(StreamServiceGrpc.stub)
  .rpc(service => { _: StreamReq =>
    val p = Promise[StreamRes]()
    service.s(new ClientResponseObserver[StreamReq, StreamRes] {
      var requestStream: ClientCallStreamObserver[StreamReq] = _

      override def beforeStart(requestStream: ClientCallStreamObserver[StreamReq]): Unit = {
        this.requestStream = requestStream
      }

      override def onNext(value: StreamRes): Unit = {
        p.trySuccess(value)
        requestStream.cancel("bye", null)
      }

      override def onError(t: Throwable): Unit = {
        p.tryFailure(t)
      }

      override def onCompleted(): Unit = {}
    })
    p.future
  })(StreamReq())

**Comment by phiSgr** [comment]
> can I use this code to do bidirectional streaming RPC?
> 
> ```
> grpc("Stream Example")
> ...
> ```

No, it won't compile.
You can certainly adapt it, but how depends on the behaviour wanted.

**Comment by ddemonaz** [comment]
Hi, I wrote such code for a bi-directional call on scala. 
```
val finishLatch = new CountDownLatch(1)
  val channel = ManagedChannelBuilder.forAddress(baseUrl, 443).build()
  val client: SpeechToTextGrpc.SpeechToTextStub = SpeechToTextGrpc.stub(channel)
  val headers = new Metadata()
  headers.put(Metadata.Key.of("authorization", Metadata.ASCII_STRING_MARSHALLER), GenerateJWT.getToken())
  val channelCustomHeader: SpeechToTextGrpc.SpeechToTextStub = MetadataUtils.attachHeaders(client, headers)

  val requestObserver: StreamObserver[StreamingRecognizeRequest] =
  channelCustomHeader.streamingRecognize(new StreamObserver[StreamingRecognizeResponse] {
    override def onNext(value: StreamingRecognizeResponse): Unit = {
      for(i <- value.results){
        for(j <- i.recognitionResult){
          println(s"Channel ${j.channel}")
          println(s"Phrase start ${j.startTime.get.seconds.toString}")
          println(s"Phrase end ${j.endTime.get.seconds},${j.endTime.get.nanos./(100000000)}")
          println(s"Is final ${i.isFinal}")
          println(s"Transcription ${j.alternatives.head.confidence}")
          println(s"Confidence ${j.alternatives.head.transcript}")
          println("----------")
        }
      }
    }
    override def onError(t: Throwable): Unit = {
      finishLatch.countDown()
      println(s"On error $t")
    }
    override def onCompleted(): Unit = {
      finishLatch.countDown()
      println("On completed")
    }
  })

  val streamingConfig: StreamingRecognitionConfig = generateAudioConfigFor_gRPC.getStreamingAudioConfig()

  val requests = List(
    StreamingRecognizeRequest()
      .withStreamingConfig(streamingConfig)) ++
    byteArray
      .grouped(1024)
      .map(x => StreamingRecognizeRequest().withAudioContent(ByteString.copyFrom(x)))
      .toList

  requests.map { x =>
     Try(requestObserver.onNext(x)).recover{
      case exception => requestObserver.onError(exception)
    }
  }

  requestObserver.onCompleted()
```

I tried to insert this code into your library based on the unary call example, but nothing works. Could you tell me how to do this? thanks

**Comment by unludo** [comment]
Hi, 
I have a bidir stream that I have to handle otherwise my whole gatling test is of no use.

I don't absolutely need to measure times for the asynchronous calls, I would just need the functionality of:
1/ initiate the bidir stream
2/ send an async message as soon as the stream is initiated
3/ receive async messages from server (which is already working in server to client stream alone)

Is it possible to achieve this with the current version of your lib?

thanks

**Comment by phiSgr** [comment]
> I don't absolutely need to measure times for the asynchronous calls, I would just need the functionality of:
> 1/ initiate the bidir stream
> 2/ send an async message as soon as the stream is initiated
> 3/ receive async messages from server (which is already working in server to client stream alone)

Not in `0.8.0`.
But in the previous versions it is possible to provide an arbitrary function that returns a `Future`. With that escape hatch you can insert whatever code you want there.

**Comment by unludo** [comment]
yes I was able to make it work. I was about to remove my post.
I understand that 0.8.0 won allow this anymore, what are the main changes of this version?
Any plan to support some bidir in the future?
thanks again for your quick response.

**Comment by phiSgr** [comment]
Removing the escape hatch allowed the `Action` to control more of the call. So it is a step to get closer to supporting streaming properly. That said, I have no idea when the next steps will be taken.
Maybe it's a good idea to put your snippets that made it work as a [gist](https://gist.github.com/), I guess it will be useful for quite a while.

`0.8.0` is quite a big release, with intertwined changes. You can see the [change log](https://github.com/phiSgr/gatling-grpc/blob/master/CHANGELOG.md) for more details.

**Comment by unludo** [comment]
ok I created a [gist](https://gist.github.com/unludo/299ddd977fb92699c233529877e97565) if it can help others.

I missed the changed log. Thanks!

**Comment by alsofr** [comment]
Hi, would it be possible to post a gist on how to read streaming response with the version 0.8.1.
The `service` method has been removed so the gist above isn't usable anymore.

**Comment by rmewanou** [comment]
For now, I reverted to use 0.7.0 in order to user the service method. Can you summarize steps needed to use streaming with 0.8.1 and we can contribute back with a PR if needed?

**Comment by alsofr** [comment]
I created this PR a few days ago as a proof of concept.
https://github.com/phiSgr/gatling-grpc/pull/27

I won't take much more to combine the result of all the responses into one grpc response.

However that won't scale to Bi directional calls as @phiSgr would like.

**Comment by rmewanou** [comment]
I actually saw that PR right after commenting on this issue, thanks @alsofr; I am looking only for server stream response processing as of now, so that PR might be enough for what I need right now. I'll give it a try and add comments if need be. Thanks for pushing it out.

**Comment by rmewanou** [comment]
FYI, I got PR #27 from @alsofr  integrated and working great with my project for server streaming calls on 0.8.2. Thanks.

**Comment by bwalks** [comment]
@unludo thanks for the gist! I am trying to use that example in a multi-step scenario where the stream part is supposed to last 60 seconds, but I keep seeing the request ending right after its created. My code is:
```
val doStream = during(60 seconds) {
  exec(startStream) // startStream is very similar to the gist example
}

val scn = scenario("Test Stream")
  .feed(data)
  .exec(createEntry)
  .exec(doStream)
  .exec(deleteEntry)
```

Any idea what I am not doing correctly?

**Comment by ddemonaz** [comment]
@phiSgr 
Hi, I implemented streaming from the client, streaming from the server and bi-directional stream. I propose to do PR in your repository. What do you think?

**Comment by phiSgr** [comment]
Thanks for the interest from all of you. Feel free to send a PR, but with a caveat.

If the way it is implemented fundamentally blocks some possible future features, I will not merge it to master. But of course I can upload that to maven central if it will be useful to others.

**Comment by UserNoOne** [comment]
Hi @alsofr .

i pulled your branch that had support for server streaming, In first go it looks good but as you mentioned currently it is not combining the result of all the responses into one gRPC response. By any chance is it ready by now ? (@rmewanou did you succeeded on combining all the responses to one ? If yes please share) 

FYI, i don't need Bi-directional streaming just a server streaming is enough

Thanks

**Comment by rmewanou** [comment]
FYI, I didn't need to combine all responses in one @UserNoOne; I simply needed the overall response time and the server streaming capability to work.

**Comment by UserNoOne** [comment]
@phiSgr @unludo After reverting gatling-grpc-plugin version to "0.7.0",  I have added a non-blocking server Stream call but seeing an issue randomly 

```
exec(grpc("Server Stream Request")
      .service(ServerStreamerGrpc.stub)
      .rpc(service => { req: StreamRequest =>
        val p = Promise[StreamResponse]()
        val responseObserver = new StreamObserver[StreamRequest ]{
          
         override def onNext(value: StreamResponse): Unit = {
            println(s"[${dT.getTime()}]==================================================>onNext ")
            p.trySuccess(value)
          }

          override def onError(throwable: Throwable): Unit = {
            println(s"[${dT.getTime()}]==================>onError!")
            if (p != null) {
              p.tryFailure(throwable)
            }
          }

          override def onCompleted(): Unit = {
            println(s"[${dT.getTime()}]==================>onCompleted!")
          }
        }
        service.serverStream(req, responseObserver)
        p.future
      })(StreamRequest .defaultInstance.withRequestId(10L))
```

**i.e** Whenever initial/first response from sever stream is **grpc-status: 0, content-length: 0, endStream=true**, gatling actor is hung up indefinitely and simulation is not being processed further

```
0:09:48.562 [DEBUG] i.g.n.NettyClientHandler - [id: 0xdbd69cd2, OUTBOUND SETTINGS: ack=false settings={ENABLE_PUSH=0, MAX_CONCURRENT_STREAMS=0, INITIAL_WINDOW_SIZE=1048576, MAX_HEADER_LIST_SIZE=8192}

10:09:48.565 [DEBUG] i.g.n.NettyClientHandler - [id: 0xdbd69cd2, OUTBOUND WINDOW_UPDATE: streamId=0 windowSizeIncrement=983041

10:09:48.574 [DEBUG] i.g.n.NettyClientHandler - [id: 0xdbd69cd2, INBOUND SETTINGS: ack=false settings={MAX_CONCURRENT_STREAMS=250, INITIAL_WINDOW_SIZE=1048576, MAX_FRAME_SIZE=1048576, MAX_HEADER_LIST_SIZE=1048896}

10:09:48.576 [DEBUG] i.g.n.NettyClientHandler - [id: 0xdbd69cd2, OUTBOUND SETTINGS: ack=true

10:09:48.609 [DEBUG] i.g.n.NettyClientHandler - [id: 0xdbd69cd2, OUTBOUND HEADERS: streamId=3 headers=GrpcHttp2OutboundHeaders[:authority: example.com:443, :path: /ServerStreamer/serverStream, :method: POST, :scheme: https, content-type: application/grpc, te: trailers, user-agent: grpc-java-netty/1.29.0, grpc-accept-encoding: gzip] streamDependency=0 weight=16 exclusive=false padding=0 endStream=false

10:09:48.617 [DEBUG] i.g.n.NettyClientHandler - [id: 0xdbd69cd2, OUTBOUND DATA: streamId=3 padding=0 endStream=true length=7 bytes=00000000020809

10:09:48.881 [DEBUG] i.g.n.NettyClientHandler - [id: 0xdbd69cd2, INBOUND WINDOW_UPDATE: streamId=0 windowSizeIncrement=983041

10:09:48.883 [DEBUG] i.g.n.NettyClientHandler - [id: 0xdbd69cd2, INBOUND SETTINGS: ack=true

10:09:48.931 [DEBUG] i.g.n.NettyClientHandler - [id: 0xdbd69cd2, INBOUND WINDOW_UPDATE: streamId=0 windowSizeIncrement=7

10:09:48.948 [DEBUG] i.g.n.NettyClientHandler - [id: 0xdbd69cd2, INBOUND HEADERS: streamId=3 headers=GrpcHttp2ResponseHeaders[:status: 200, content-type: application/grpc, grpc-status: 0, content-length: 0, date: Wed, 23 Sep 2020 04:39:48 GMT] padding=0 endStream=true

[1600835986150]==================>onCompleted!

```

**Comment by UserNoOne** [comment]
Hello people,

Above issue is resolved! 
if the server responds with a zero-length stream (no messages), then onNext() won't be called and **"onCompleted"** flag is still false in this case. So, explicitly setting the onCompleted flag in "onCompleted" callback solved the issue

```
override def onCompleted(): Unit = {
    if(!promise.isCompleted) {
       println("Server responds with a zero-length stream, So setting isCompleted Flag explicitly")
       p.trySuccess(new StreamServerResponse)
   }
} 
```

**Comment by phiSgr** [comment]
Closed with b164e57

Took me two years lol.
